src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java

```diff
@@ -78,46 +95,47 @@ public ProductReferenceResolver(@Nonnull final ProductSyncOptions productSyncOpt
+        final ProductVariantDraft productDraftMasterVariant = draftBuilder.getMasterVariant();
         if (productDraftMasterVariant != null) {
             return resolveProductVariantPriceReferences(productDraftMasterVariant)
-                .thenApply(resolvedMasterVariant ->
-                    ProductDraftBuilder.of(productDraft)
-                                       .masterVariant(resolvedMasterVariant).build())
+                .thenApply(draftBuilder::masterVariant)
                 .thenCompose(this::resolveProductVariantsPriceReferences);
         }
-        return resolveProductVariantsPriceReferences(productDraft);
+        return resolveProductVariantsPriceReferences(draftBuilder);
     }
 
     @Nonnull
-    private CompletionStage<ProductDraft> resolveProductVariantsPriceReferences(
-        @Nonnull final ProductDraft productDraft) {
-        final List<ProductVariantDraft> productDraftVariants = productDraft.getVariants();
+    private CompletionStage<ProductDraftBuilder> resolveProductVariantsPriceReferences(
+        @Nonnull final ProductDraftBuilder draftBuilder) {
+        final List<ProductVariantDraft> productDraftVariants = draftBuilder.getVariants();
         if (productDraftVariants == null) {
-            return CompletableFuture.completedFuture(productDraft);
+            return CompletableFuture.completedFuture(draftBuilder);
         }
 
         final List<CompletableFuture<ProductVariantDraft>> resolvedVariantFutures =
             productDraftVariants.stream()
-                                .filter(Objects::nonNull)
-                                .map(this::resolveProductVariantPriceReferences)
-                                .map(CompletionStage::toCompletableFuture)
-                                .collect(Collectors.toList());
+                .filter(Objects::nonNull)
+                .map(this::resolveProductVariantPriceReferences)
+                .map(CompletionStage::toCompletableFuture)
+                .collect(Collectors.toList());
         return
             CompletableFuture.allOf(
                 resolvedVariantFutures.toArray(new CompletableFuture[resolvedVariantFutures.size()]))
-                             .thenApply(result -> resolvedVariantFutures.stream()
-                                                                        .map(CompletableFuture::join)
-                                                                        .collect(Collectors.toList()))
-                             .thenApply(resolvedVariants ->
-                                 ProductDraftBuilder.of(productDraft).variants(resolvedVariants).build());
+                .thenApply(result -> resolvedVariantFutures.stream()
+                    .map(CompletableFuture::join)
+                    .collect(Collectors.toList()))
+                .thenApply(draftBuilder::variants);
     }
 
     private CompletionStage<ProductVariantDraft> resolveProductVariantPriceReferences(
```

Comment:
Unfortunately, the `ProductDraftBuilder#variants` and `ProductDraftBuilder#masterVariant` accept only built drafts, not draft builders, so we will need explicitly build them anyway, just one step later.

Another issue: JVM SDK `ProductDraftBuilder#variants(List<ProductVariantDraft>)` does not accept wildcard lists, in the same time the `ProductVariantDraftBuilder#build` returns more specific `ProductVariantDraftDsl` which is collected later to `List<ProductVariantDraftDsl>` and thus can't be supplied to `#variants(List<ProductVariantDraft>)`
I'll supply an issue for the JVM SDK.

This is how the patch looks in this case:
(Note explicit mapping `map(draftDsl -> (ProductVariantDraft)draftDsl)` and `ProductVariantDraftBuilder::build` applied twice)
```bash
@@ -109,6 +109,7 @@ public final class ProductReferenceResolver extends BaseReferenceResolver<Produc
         final ProductVariantDraft productDraftMasterVariant = draftBuilder.getMasterVariant();
         if (productDraftMasterVariant != null) {
             return resolveProductVariantPriceReferences(productDraftMasterVariant)
+                .thenApply(ProductVariantDraftBuilder::build)
                 .thenApply(draftBuilder::masterVariant)
                 .thenCompose(this::resolveProductVariantsPriceReferences);
         }
@@ -123,29 +124,32 @@ public final class ProductReferenceResolver extends BaseReferenceResolver<Produc
             return CompletableFuture.completedFuture(draftBuilder);
         }
 
-        final List<CompletableFuture<ProductVariantDraft>> resolvedVariantFutures =
+        final List<CompletableFuture<ProductVariantDraftBuilder>> resolvedVariantFutures =
             productDraftVariants.stream()
                 .filter(Objects::nonNull)
                 .map(this::resolveProductVariantPriceReferences)
                 .map(CompletionStage::toCompletableFuture)
                 .collect(Collectors.toList());
         return
             CompletableFuture.allOf(
                 resolvedVariantFutures.toArray(new CompletableFuture[resolvedVariantFutures.size()]))
                 .thenApply(result -> resolvedVariantFutures.stream()
                     .map(CompletableFuture::join)
+                    .map(ProductVariantDraftBuilder::build)
+                    .map(draftDsl -> (ProductVariantDraft)draftDsl)
                     .collect(Collectors.toList()))
                 .thenApply(draftBuilder::variants);
     }
 
-    private CompletionStage<ProductVariantDraft> resolveProductVariantPriceReferences(
+    private CompletionStage<ProductVariantDraftBuilder> resolveProductVariantPriceReferences(
         @Nonnull final ProductVariantDraft productVariantDraft) {
         final List<PriceDraft> productVariantDraftPrices = productVariantDraft.getPrices();
         final ProductVariantDraftBuilder productVariantDraftBuilder =
             ProductVariantDraftBuilder.of(productVariantDraft);
 
         if (productVariantDraftPrices == null) {
-            return completedFuture(productVariantDraftBuilder.build());
+            return completedFuture(productVariantDraftBuilder);
         }
 
         final List<CompletableFuture<PriceDraft>> resolvedPriceDraftFutures =
@@ -158,7 +162,7 @@ public final class ProductReferenceResolver extends BaseReferenceResolver<Produc
             .thenApply(result -> resolvedPriceDraftFutures.stream()
                 .map(CompletableFuture::join)
                 .collect(Collectors.toList()))
-            .thenApply(resolvedPriceDrafts -> productVariantDraftBuilder.prices(resolvedPriceDrafts).build());
+            .thenApply(productVariantDraftBuilder::prices);
     }
 
```

Reply:
In the same time i've simplified a bit the approach (avoid redundant `toList()` collecting):
08727bbd

Reply:
> so we will need explicitly build them anyway, just one step later.
But that is what you do as well when you get a productDraft in resolveReferences:

you do: `resolveProductTypeReference(ProductDraftBuilder.of(productDraft))` which is what you will do/are doing in this patch.. (Anyways after talking, this would be changes in #160)